src/Models/BaseElement.php

```diff
@@ -712,35 +712,65 @@ public function CMSEditLink($directLink = false)
             return $this->cacheData['cms_edit_link'];
         }
 
-        $relationName = $this->getAreaRelationName();
-        $page = $this->getPage();
+        $link = $this->getElementCMSLink($directLink);
+        $this->extend('updateCMSEditLink', $link);
 
-        if (!$page) {
-            $link = null;
-            $this->extend('updateCMSEditLink', $link);
-            return $link;
+        if ($link) {
+            $this->cacheData['cms_edit_link'] = $link;
         }
 
-        if (!$page instanceof SiteTree && method_exists($page, 'CMSEditLink')) {
-            $link = Controller::join_links($page->CMSEditLink(), 'ItemEditForm');
-        } else {
-            $link = $page->CMSEditLink();
-        }
+        return $link;
+    }
 
-        // In-line editable blocks should just take you to the page. Editable ones should add the suffix for detail form
-        if (!$this->inlineEditable() || $directLink) {
-            $link = Controller::join_links(
-                singleton(CMSPageEditController::class)->Link('EditForm'),
-                $page->ID,
-                'field/' . $relationName . '/item/',
-                $this->ID,
-                'edit'
-            );
-        }
+    /**
+     * 
+     * @param bool $directLink Indicates that the GridFieldDetailEdit
+     *             form link should be given even if the block can be edited in-line.
+     * @return null|string
+     * @throws \SilverStripe\ORM\ValidationException
+     */
+    public function getElementCMSLink($directLink = false)
+    {
+        $relationName = $this->getAreaRelationName();
+        $page = $this->getPage();
 
-        $this->extend('updateCMSEditLink', $link);
+        $link = null;
+
+        if ($page) {
+            // If $page is not a Page
+            if (!($page instanceof SiteTree)) {
```

Comment:
```suggestion
            if ($page instanceof SiteTree) {
```
And then reverse the blocks below